src/ModelContextProtocol/Shared/McpSession.cs

```diff
@@ -487,77 +500,127 @@ public async Task SendMessageAsync(IJsonRpcMessage message, CancellationToken ca
-                        }
-                    }
-                    break;
+            if (message is IJsonRpcMessageWithId withId)
+            {
+                activity.AddTag("mcp.request.id", withId.Id.Id?.ToString());
             }
         }
+
+        JsonObject? paramsObj = message switch
+        {
+            JsonRpcRequest request => request.Params as JsonObject,
+            JsonRpcNotification notification => notification.Params as JsonObject,
+            _ => null
+        };
+
+        if (paramsObj == null)
+        {
+            return;
+        }
+
+        string? target = null;
+        switch (method)
+        {
+            case RequestMethods.ToolsCall:
+                string? toolName = GetStringProperty(paramsObj, "name");
+                if (toolName is not null)
+                {
+                    tags.Add("mcp.tool.name", toolName);
+                    target = toolName;
+                }
+                break;
+
+            case RequestMethods.PromptsGet:
+                string? promptName = GetStringProperty(paramsObj, "name");
+                if (promptName is not null)
+                {
+                    tags.Add("mcp.prompt.name", promptName);
+                    target = promptName;
+                }
+                break;
+
+            case RequestMethods.ResourcesRead:
+            case RequestMethods.ResourcesSubscribe:
+            case RequestMethods.ResourcesUnsubscribe:
+            case NotificationMethods.ResourceUpdatedNotification:
+                string? resourceUri = GetStringProperty(paramsObj, "uri");
+                if (resourceUri is not null)
+                {
+                    tags.Add("mcp.resource.uri", resourceUri);
+                    target = resourceUri;
+                }
+                break;
+        }
+
+        if (activity is { IsAllDataRequested: true })
+        {
+            activity.DisplayName = target == null ? method : $"{method} {target}";
+        }
     }
 
-    private static void AddExceptionTags(ref TagList tags, Exception e)
+    private static void AddExceptionTags(ref TagList tags, Activity? activity, Exception e)
     {
         if (e is AggregateException ae && ae.InnerException is not null and not AggregateException)
         {
             e = ae.InnerException;
         }
 
-        tags.Add("error.type", e.GetType().FullName);
-        tags.Add("rpc.jsonrpc.error_code",
-            (e as McpException)?.ErrorCode is int errorCode ? errorCode :
-            e is JsonException ? ErrorCodes.ParseError :
-            ErrorCodes.InternalError);
+        int? intErrorCode = (e as McpException)?.ErrorCode is int errorCode ? errorCode :
+            e is JsonException ? ErrorCodes.ParseError : null;
+
+        tags.Add("error.type", intErrorCode == null ? e.GetType().FullName : intErrorCode.ToString());
+        if (intErrorCode is not null)
+        {
+            tags.Add("rpc.jsonrpc.error_code", intErrorCode.ToString());
```

Comment:
I'm fine either way, but are you purposefully still using the jsonrpc terminology here?

Reply:
yes, I don't  want to invent a new attribute since MCP does not define this code.